
\documentclass[12pt]{article}

\usepackage{hyperref}

\newcommand{\R}[1]{{\texttt{#1}}}
\newcommand{\C}[1]{{\texttt{#1}}}

\textwidth=6.2in
\textheight=8.5in
\oddsidemargin=0.2in
\evensidemargin=0.2in
\headheight=0in
\headsep=0in
\usepackage{Sweave}

\begin{document}
\title{samExploreR package: high-performance read summarisation with sequencing depth reduction simulation}
\author{Alexey Stupnikov}
\date{23 March 2015}
\maketitle

\section{Introduction}
In this vignette we introduce the samExploreR package.

The samExploreR package allows to perform a summarisation of NGS experiment reads for given annotation with simulation of reduced sequencing depth.
During the summarisation process a fraction of reads is removed to simulate an NGS experiment with smaller library size.




\section{Depth simulation}

Function \R{samExplore} performs reads summarisation similar to \R{featureCounts} function from Rsubread package.

To simulate a sequencing depth, for given $f$ that may be variated from $0$ to $1$ during the summarisation process a read is counted or not with probability $f$.
This process is repeated $N$ times.

\R{samExplore} takes an input list if SAM or BAM files containing mapped reads, annotation file with genomic features such as genes and exons to assign to and several perfomance parameters: some of them are the ones that are passed to \R{featureCounts} function and in addition fraction of reads to remove $f$ and number of repeats $N$


%\begin{scriptsize}
%<<>>=
%## Fraction of A,T,G and C bases in all the reads
%x <- atgcContent(filename=reads)
%x
%## Fraction of A,T,G and C bases at each read location
%xb <- atgcContent(filename=reads,basewise=TRUE)
%xb[,1:5]
%@
%\end{scriptsize}


\section{Citation}

\section{Authors}
Alexey Stupnikov. Shailesh Tripathi, Frank Emmert-Streib \\


\section{Contact}


\end{document}



